### PR TITLE
.github/workflows: build NixOS/Nixpkgs manuals 

### PR DESCRIPTION
**Pull request by zowoq**
Continued from https://github.com/NixOS/nixpkgs/pull/87853

cc @grahamc @Mic92 

Using `ofborg` is probably better but this might be useful for now?

This action will only start if something in `/doc` or `/nixos/doc` is modified so most contributors shouldn't see it.

Each job will skip the cachix action/building the manual unless its paths have changed. 

They could be split further into one action per path if we really wanted to save a couple of seconds.

### CODE REVIEW

File: .github/workflows/manual.yml
Comment (Mic92) [inline .github/workflows/manual.yml:7]: Should we also include `nixos/modules/**`? I had a few cases where module option description broke the manual. Of all PRs we have only a few are touching modules.

File: .github/workflows/manual.yml
Comment (Mic92) [inline .github/workflows/manual.yml:26]: According to the manual https://nixos.org/nixpkgs/manual/#chap-contributing we build it with `nix-build ./doc`. However it also says that `make -C doc debug` might come in handy in case there are errors.

Comment (Mic92) [reply .github/workflows/manual.yml:26]: cc @grahamc

Comment (zowoq) [reply .github/workflows/manual.yml:26]: I hadn't noticed that `ofborg` is running [instantiate](https://github.com/NixOS/ofborg/blob/released/ofborg/src/tasks/eval/nixpkgs.rs#L480-L502) for the manuals, `nixpkgs` has `pkgs/top-level/release.nix -A manual`.

**Comment by zowoq** [comment]
The outcome of the doc format RFC might mean that this isn't needed in a few months which might actually be a point in its favour as it can be easily dropped. 

I'm assuming that adding these builds to `ofborg` is non-trivial and wouldn't be a productive use of time if we might end up changing formats anyway.

**Comment by Mic92** [comment]
I also like github workflows for these sort of task more since they don't involve writing Rust and are easier to prototype.

**Comment by grahamc** [comment]
I like it too. I don’t think this is a good idea, because they will fail during mass rebuilds from being too big. We don’t want to train users that red X’s can be ignored.

**Comment by grahamc** [comment]
cross-posted from #87853:

I would like to revert this PR and close #88488. Not because I don't like github actions or feel like my pet project ofborg is threatened, but because I think these checks are actually harming the clearness of the checks. I'll explain:

The editorconfig check puts a big green check next to a commit, and hides the list of checks which passed. If ofborg is down or behind, it is likely people will think ofborg has approved the PR before it did. This could cause, well, all the problems ofborg is there to prevent.

The concrete problem here is this green check appears very early and is misleading, and there is no poka yoke to help people do the right thing.

One fear of #88488 is that it will introduce red X's if it takes a long time to build the manual, or something in the target branch is failing to build. This is reminiscent of the Travis times for me: Travis gave out so few green checkmarks, that if Travis gave a PR a green check-mark, it inspired more "okay what is broken that let travis pass it?" than "okay good to merge".
 
The goal of ofborg, and ever since ofborg became a thing, the "rule" has been green checkmark -> good to merge and red X -> don't merge. 

The editorconfig check errodes the clearness of the green checkmark signal.

The big-rebuild failures we'll certainly have with #88488 errodes the clearness of the red X signal.

The clearness of this signal has been my guiding light on the design of ofborg for years now, and I am very afraid of threatening it. 

A few suggestions that have been brought up to improve this:

1. **make ofborg faster** I'd love to do this, but it doesn't solve this problem if ofborg is seriously delayed or down.
2. **require the ofborg evaluation check for merges** this is really the only solution, but not something we can do today, because of some workflow issues we need to resolve.

That makes me think that, today, we should be undoing this check and closing #88488.

**Comment by Mic92** [comment]
We can now safely run github actions since there is an action that just waits for ofborg. However is this PR here still needed since ofborg already builds the manual?

**Comment by zowoq** [comment]
> However is this PR here still needed since ofborg already builds the manual?

AFAIK `ofborg` doesn't?

With the docs format changes ongoing I wasn't planning on doing any work on this for now.

**Comment by Mic92** [comment]
ok.

**Comment by zowoq** [comment]
Hmm, I guess the actual docs format change may not happen for a while but in the meantime we still have broken builds.

May as well give this a go.

**Comment by zowoq** [comment]
Now that the install-nix-action is much faster using the full clone / git diff action is the bottleneck. May as well just run the build anyway.

**Comment by zowoq** [comment]
Binary cache was mentioned in the other builds PR, would running the manual builds on master as an hourly cron job and pushing to the cache address the rebuild concerns?

**Comment by Mic92** [comment]
> Binary cache was mentioned in the other builds PR, would running the manual builds on master and staging as an hourly cron job and pushing to the cache address the rebuild concerns?

This would not be even necessary. We could a cachix signing key that would just upload the manual on every push to master.

**Comment by Mic92** [comment]
> > Binary cache was mentioned in the other builds PR, would running the manual builds on master and staging as an hourly cron job and pushing to the cache address the rebuild concerns?
> 
> This would not be even necessary. We could a cachix signing key that would just upload the manual on every push to master.

There might be one problem with approach though:

https://github.com/NixOS/nixpkgs/pull/94658#issuecomment-669768239

**Comment by Mic92** [comment]
That might also work: https://docs.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows

File: .github/workflows/manual.yml
Comment (Mic92) [inline .github/workflows/manual.yml:10]: We have two bots @ofborg and @marvin-mk2. Marvin-mk2 also uses the `/` syntax. Probably we should just stick to that. cc @timokau

**Review by Mic92 — APPROVED** [review]
I am in favor of running this on-demand.

**Comment by Mic92** [comment]
I think actions might not run when in draft mode?

**Comment by zowoq** [comment]
> https://docs.github.com/en/actions/reference/events-that-trigger-workflows#issue_comment
>
> This event will only trigger a workflow run if the workflow file is on the master or default branch.

**Comment by Mic92** [comment]
> > [docs.github.com/en/actions/reference/events-that-trigger-workflows#issue_comment](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#issue_comment)
> > This event will only trigger a workflow run if the workflow file is on the master or default branch.

Does it also use the execution context of master? That would be great because in that case we could use it for our cachix builds.

**Comment by zowoq** [comment]
Looks like if we want a job status we need to set it manually or just post a comment when it finishes with pass/fail and a link, would probably need to set another token for it to work.

File: .github/workflows/manual.yml
Comment (timokau) [reply .github/workflows/manual.yml:10]: You mean stick to the syntax? I picked it for marvin since github doesn't autocomplete @mentions for gh-apps. Note that there are a few other events too that could contain the text (pull request reviews etc) if you care for completeness. Just issue comments is probably good enough for this purpose though.

**Comment by zowoq** [comment]
I've switched this back to running when paths are modified and to use `pull_request_target`.

@Mic92 Which cachix account would we wan to use, `nix-community` or something else?

**Comment by Mic92** [comment]
> I've switched this back to running when paths are modified and to use `pull_request_target`.
> 
> @Mic92 Which cachix account would we wan to use, `nix-community` or something else?

This is up to @domenkozar

**Comment by domenkozar** [comment]
I think it's best to create a separate account. Not sure who should do that, probably someone from the infrastructure team like @zimbatm

**Comment by zowoq** [comment]
I might just create a cache under my name and include the key as plaintext in the workflow for now, we can switch to an official account later if this works.

**Comment by Mic92** [comment]
@zowoq ok, but you likely need to give the credentials to someone who has administrative rights in nixpkgs like @domenkozar so its accessible from the NixOS org when starting actions from master.

**Comment by zimbatm** [comment]
@zowoq I added a signing key to the repo. I think it's fine to merge now.

**Comment by zowoq** [comment]
Thanks @zimbatm.

I've split this into two separate actions so that the entire job is conditional on its path being modified.

**Review by Mic92 — APPROVED** [review]

**Comment by zowoq** [comment]
Apologies for the belated follow up on this, I reverted this shortly after it was merged as it wasn't working correctly and forgot to comment here.

`pull_request_target` was checking out only the base branch and not the PR branch so it wasn't actually doing anything except verifying that the manuals were building on the base branch. `pull_request_target` also allows write access to the repo so we can't just force it to check out the PR branch.

`pull_request` doesn't have write access but it will expose the cache token which is what we were trying to avoid, IIRC there have been some changes on the Cachix side with tokens but I haven't investigated that yet to see if it would help here.

**Comment by Mic92** [comment]
> Apologies for the belated follow up on this, I reverted this shortly after it was merged as it wasn't working correctly and forgot to comment here.
> 
> `pull_request_target` was checking out only the base branch and not the PR branch so it wasn't actually doing anything except verifying that the manuals were building on the base branch. `pull_request_target` also allows write access to the repo so we can't just force it to check out the PR branch.

Can it not checkout the pull request branch without executing code from it besides building the manual?

**Comment by zowoq** [comment]
Maybe I'm misunderstanding what you mean but wouldn't building the manual also build any of it's dependenices that the PR modifies anyway?

**Comment by Mic92** [comment]
> Maybe I'm misunderstanding what you mean but wouldn't building the manual also build any of it's dependenices that the PR modifies anyway?

Yes, but in a nix sandbox, no? The nix binary would be provided by the installer-action. I think the most important security aspect would be that no secrets are in environment variables and than the nix sandbox should take care of the rest.

**Comment by zowoq** [comment]
>  I think the most important security aspect would be that no secrets are in environment variables 

What about the cachix token?

**Comment by Mic92** [comment]
> > I think the most important security aspect would be that no secrets are in environment variables
> 
> What about the cachix token?

I think it no longer is exported via environment variables. I think the `with` statement only sets it for the cachix action itself: https://nix.dev/tutorials/continuous-integration-github-actions.html
Some sort of restricted eval to include files from the systemd should be disabled so.

**Comment by zowoq** [comment]
Let's try this again: https://github.com/NixOS/nixpkgs/pull/106036

**Pull request merged**
